### DISCUSSION TITLE
Does `SecurityIdentityAssociation` have to be `@RequestScoped`?

### DISCUSSION THREAD

**Discussion by knutwannheden** [quarkusio/quarkus#28742] [Q&A] [answered]
`SecurityIdentityAssociation` is the implementation of `CurrentIdentityAssociation` and is declared as `@RequestScoped`. I was wondering if it actually has to be request scoped or whether it should instead be implemented using Vert.x context locals.

My background is that we are doing processing using a scheduled method (using `@Scheduled`), where we had problems with OpenTelemetry. This problem was addressed by making sure the scheduled method is executed on a Vert.x duplicated context (see https://github.com/quarkusio/quarkus/pull/28255).

Further, we are also processing incoming Kafka messages, where we sometimes want to set a Quarkus security identity. Currently, this requires that the `@Incoming`-annotated method is also annotated with `@ActivateRequestContext`.

[a comment by an automated account is not shown]

**Reply by sberyozkin** [reply]
Hi @knutwannheden I'm not sure I understand the consequences, not sure such a refactoring is necessary given that adding `@ActivateRequestContext`, even though it requires some effort, works...

**Reply by knutwannheden** [reply]
Adding the annotation isn't that much work, it may just be a bit of a surprise for the user, since the concept of requests doesn't really exist in that sense when processing Kafka messages.

I was just thinking the decision to use `@RequestScoped` for `SecurityIdentityAssociation` may have been a decision from the early days, which maybe could benefit from being revisited.

Pinging @cescoffier here, since he worked on #28255.

**Comment by sberyozkin** [comment, marked as answer]
The thing is that SecuriryIdentity represents the current user credentials in the current HTTP request. I'm not sure changing these semantics to make them available in the duplicated contexts during the request which will happen at some time later is a good idea right now... It is not even the same user interaction

**Reply by knutwannheden** [reply]
Thanks for the explanation. I still haven't quite grasped how exactly duplicated contexts are used and propagated in reactive requests and how they are used for OpenTelemetry. I was just wondering if it would potentially simplify things to also use them for the security context.

**Reply by cescoffier** [reply]
@sberyozkin isn't the security context stored in the request context? If so... it's stored in the duplicated context.

**Reply by sberyozkin** [reply]
But would we like to have this security context available to the out of band/unrelated requests like Kafka messages, that is the concern I have...

**Reply by cescoffier** [reply]
In general you don't want to do that. First, there is a security concern. Then, token are likely expired when the record is processed. You can attach the principal/user to the record if you want (as header, or as part of the key/value pair)

**Reply by knutwannheden** [reply]
Also, in relation to this, I have noticed that Quarkus has no explicit "implementation" of the MicroProfile `ThreadContext.SECURITY` context. I assume that is because it is implemented using the CDI request scope and as such can't really be separated from the CDI scope and thus also can't be propagated (or cleared) independently. (The same basically applies to the OpenTelemetry tracing context, but that is a different issue.)

Maybe supporting MicroProfile `ThreadContext.SECURITY` would provide a way forward for use cases when the user explicitly wants to propagate that context independently of the request context.

**Reply by sberyozkin** [reply]
Hi @knutwannheden, Can you please open an enhancement request, for MP ContextPropagation impl to support this type of context ? I don't know how feasible it will be to support it, but tracking it with a dedicated issue would make it move visible

**Reply by knutwannheden** [reply]
@sberyozkin AFAICT the MP Context Propagation specification already supports this. I think there is just the implementation missing in Quarkus or am I missing something?

**Reply by sberyozkin** [reply]
@knutwannheden This is what I meant, sorry, have a Quarkus issue, `for MP ContextPropagation impl to support this type of context`, it was a bit cryptic I agree :-), `impl` there means Quarkus implementation of MP Context Propagation

**Reply by knutwannheden** [reply]
See #29703.
